src/Polysemy/Fresh.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE BangPatterns, TemplateHaskell, Trustworthy #-}
+module Polysemy.Fresh
+  (-- * Effect
+    Fresh(..)
+
+    -- * Actions
+  , fresh
+
+    -- * Interpretations
+  , freshToIO
+
+    -- * Unsafe Interpretations
+  , runFreshUnsafe
+  , runFreshUnsafePerformIO
+  ) where
+
+import Data.Unique
+
+import Polysemy.Internal
+import Polysemy.Internal.Union
+import System.IO.Unsafe (unsafePerformIO)
+import Polysemy
+import Polysemy.State
+
+-----------------------------------------------------------------------------
+-- | An effect for creating unique objects which may be used as references,
+-- a la 'Unique'. Polymorphic code making use of 'Fresh' is expected
+-- to place constraints upon @uniq@ as necessary.
+--
+-- Any interpreter for 'Fresh' has the responsibilty of ensuring
+-- that any call to 'fresh' produces an object that /never/
+-- compares equal to an object produced by a previous call to 'fresh'.
+data Fresh uniq m a where
+  Fresh :: Fresh uniq m uniq
```

Comment:
I'm deviating from `fused-effects` here in that I don't provide a `reset` action. My intention with this effect is that it is a unique symbol generator; nothing more, and definitely nothing less. Having `reset` would contradict that, and is also the reason why the pure interpreter (`runFreshUnsafe`) is marked as unsafe, since the state could get backtracked by another effect/the final monad, so `fresh` in those cases doesn't guarantee that the symbol it generates is unique.

Reply:
I'm cool with this. That's all I ever use it for too.